nemo/collections/nlp/data/language_modeling/megatron/data_samplers.py

```diff
@@ -149,9 +150,13 @@ def __init__(
             pad_samples_to_global_batch_size == False
         ), "`MegatronPretrainingRandomSampler` does not support sample padding"
         self.last_batch_size = self.total_samples % self.micro_batch_times_data_parallel_size
+        self.seed = seed
 
     def __len__(self):
-        num_available_samples: int = self.total_samples
+        active_total_samples = self.total_samples - (self.last_batch_size if self.drop_last else 0)
```

Comment:
Probably that the same formula should be used in `__iter__()`?

Edit: although I'm afraid this would trigger the `assert current_epoch_samples % self.micro_batch_times_data_parallel_size == 0`... what about just adding a comment in `__iter__()` that `self.epoch` may be incorrect when `drop_last` is False (and not try to fix it in this PR unless you're confident in the fix, since this bug already exists anyway).

Reply:
I think that the way these samplers are built to work (sharding out to each DP), they can never work correctly with `drop_last=False` because then you do not have an equivalent shard length on each DP which will cause a hang. `drop_last=False` would only ever work correctly if DP=1.

Perhaps we should add an assert in the constructor which would only ever allow `drop_last=False` if and only if `data_parallel_size == 1`?

Reply:
I have added a new raise exception which ensures nobody can use the class with `drop_last=False` and MBS * DP > 1. I think this solves all issues and circumvents the necessity of adding a comment under `__iter__`

Reply:
From where I stand I agree with you, but I'm not sure exactly how these samplers are used (by other code / other users) so I'm afraid I can't really confirm whether it's ok to raise an exception in such a situation.